File: BST/AVL/lingkerio/OCaml-version/AVLTree/lib/avl.ml
(* lib/avl.ml *)

(* Definition type for an AVL tree *)
type 'a avl =
  | Empty
  | Node of 'a * 'a avl * 'a avl * int

(* Create an empty AVL tree *)
let empty = Empty

(* Get the height of an AVL tree node *)
let height = function
  | Empty -> 0
  | Node (_, _, _, h) -> h

(* Find the minimum value in the AVL tree *)
let rec min_avl = function
  | Empty -> None
  | Node (x, Empty, _, _) -> Some x
  | Node (_, left, _, _) -> min_avl left

(* Find the maximum value in the AVL tree *)
let rec max_avl = function
  | Empty -> None
  | Node (x, _, Empty, _) -> Some x
  | Node (_, _, right, _) -> max_avl right

(* Check if a value exists in the AVL tree *)
let rec member x = function
  | Empty -> false
  | Node (y, left, right, _) ->
    if x < y then member x left
    else if x > y then member x right
    else true

(* Calculate the balance factor of a node *)
let balance_factor = function
  | Empty -> 0
  | Node (_, left, right, _) -> height left - height right

(* Rotate left *)
let rotate_left = function
  | Node (x, a, Node (y, b, c, _), _) ->
    Node (y, Node (x, a, b, 1 + max (height a) (height b)),
               c, 1 + max (1 + max (height a) (height b)) (height c))
  | _ -> failwith "Invalid rotation"

(* Rotate right *)
let rotate_right = function
  | Node (y, Node (x, a, b, _), c, _) ->
    Node (x, a, Node (y, b, c, 1 + max (height b) (height c)),
               1 + max (height a) (1 + max (height b) (height c)))
  | _ -> failwith "Invalid rotation"

(* Rotate left-right *)
let rotate_left_right = function
  | Node (x, a, Node (y, b, c, _), _) ->
    rotate_left (Node (x, a, rotate_right (Node (y, b, c, 1 + max (height b) (height c))),
                             1 + max (height a) (1 + max (height b) (height c))))
  | _ -> failwith "Invalid rotation"

(* Rotate right-left *)
let rotate_right_left = function
  | Node (y, Node (x, a, b, _), c, _) ->
    rotate_right (Node (y, rotate_left (Node (x, a, b, 1 + max (height a) (height b))),
                             c, 1 + max (1 + max (height a) (height b)) (height c)))
  | _ -> failwith "Invalid rotation"

(* Balance the AVL tree *)
let balance = function
  | Node (x, a, b, _) when balance_factor (Node (x, a, b, height a)) > 1 ->
    if balance_factor a >= 0 then rotate_right (Node (x, a, b, 1 + max (height a) (height b)))
    else rotate_right_left (Node (x, a, b, 1 + max (height a) (height b)))
  | Node (x, a, b, _) when balance_factor (Node (x, a, b, height b)) < -1 ->
    if balance_factor b <= 0 then rotate_left (Node (x, a, b, 1 + max (height a) (height b)))
    else rotate_left_right (Node (x, a, b, 1 + max (height a) (height b)))
  | node -> node

(* Insert a value into the AVL tree *)
let rec insert x = function
  | Empty -> Node (x, Empty, Empty, 1)
  | Node (y, left, right, _) as node ->
    let new_node = 
      if x < y then
        let new_left = insert x left in
        Node (y, new_left, right, 1 + max (height new_left) (height right))
      else if x > y then
        let new_right = insert x right in
        Node (y, left, new_right, 1 + max (height left) (height new_right))
      else node
    in
    balance new_node

(* Delete a value from the AVL tree *)
let rec delete x = function
  | Empty -> Empty
  | Node (y, left, right, _) ->
    let new_node = 
      if x < y then
        let new_left = delete x left in
        Node (y, new_left, right, 1 + max (height new_left) (height right))
      else if x > y then
        let new_right = delete x right in
        Node (y, left, new_right, 1 + max (height left) (height new_right))
      else
        match left, right with
        | Empty, _ -> right
        | _, Empty -> left
        | _ ->
          (match min_avl right with
          | Some min_right ->
            let new_right = delete min_right right in
            Node (min_right, left, new_right, 1 + max (height left) (height new_right))
          | None -> failwith "Unexpected None in delete")
    in
    balance new_node

File: BST/AVL/wisdomgo/cpp-version/src/avl.cpp
#include "avl.h"
#include <algorithm>
#include <stdexcept>

AVL::AVL() : root(nullptr) {}

AVL::~AVL() {
    clear(root);
}

void AVL::clear(AVLnode* node) {
    if (node) {
        clear(node->left);
        clear(node->right);
        delete node;
    }
}

int height(AVLnode* node) {
    return node ? node->height : 0;
}

int getBalance(AVLnode* node) {
    if (!node) return 0;
    return height(node->left) - height(node->right);
}

AVLnode* rightRotate(AVLnode* y) {
    AVLnode* x = y->left;
    AVLnode* T2 = x->right;
    x->right = y;
    y->left = T2;
    y->height = std::max(height(y->left), height(y->right)) + 1;
    x->height = std::max(height(x->left), height(x->right)) + 1;
    return x;
}

AVLnode* leftRotate(AVLnode* x) {
    AVLnode* y = x->right;
    AVLnode* T2 = y->left;
    y->left = x;
    x->right = T2;
    x->height = std::max(height(x->left), height(x->right)) + 1;
    y->height = std::max(height(y->left), height(y->right)) + 1;
    return y;
}

AVLnode* AVL::insert(AVLnode* node, int val) {
    if (!node) return new AVLnode(val);
    if (val < node->val)
        node->left = insert(node->left, val);
    else if (val > node->val)
        node->right = insert(node->right, val);
    else
        return node;
    node->height = 1 + std::max(height(node->left), height(node->right));
    int balance = getBalance(node);
    if (balance > 1 && val < node->left->val) // LL型，右旋
        return rightRotate(node);
    if (balance < -1 && val > node->right->val)
        return leftRotate(node);  // RR型，左旋
    if (balance > 1 && val > node->left->val) {
        node->left = leftRotate(node->left);
        return rightRotate(node);   //LR型，先左旋再右旋
    }
    if (balance < -1 && val < node->right->val) {
        node->right = rightRotate(node->right);
        return leftRotate(node);  // RL型，先右旋再左旋
    }

    return node;
}
AVLnode* AVL::findmin(AVLnode* node) {
    if (node == nullptr) return nullptr;
    while (node->left != nullptr) {
        node = node->left;
    }
    return node;
}

void AVL::insert(int val) {
    root = insert(root, val);
}

bool AVL::find(int val) {
    AVLnode* current = root;
    while (current) {
        if (val < current->val)
            current = current->left;
        else if (val > current->val)
            current = current->right;
        else
            return true; 
    }
    return false; 
}

int AVL::findmin() {
    AVLnode* current = root;
    if (!current) throw std::runtime_error("Tree is empty");
    while (current->left != nullptr)
        current = current->left;
    return current->val;
}

int AVL::findmax() {
    AVLnode* current = root;
    if (!current) throw std::runtime_error("Tree is empty");
    while (current->right != nullptr)
        current = current->right;
    return current->val;
}

AVLnode* AVL::remove(AVLnode* node, int val, bool& isDeleted) {
    if (node == nullptr) {
        isDeleted = false;
        return nullptr;
    }
    if (val < node->val) {
        node->left = remove(node->left, val, isDeleted);
    } else if (val > node->val) {
        node->right = remove(node->right, val, isDeleted);
    } else {
        if (node->left == nullptr || node->right == nullptr) {
            AVLnode* temp = node->left ? node->left : node->right;
            if (temp == nullptr) {
                delete node;
                node = nullptr;
            } else {
                *node = *temp; 
                delete temp;
            }
            isDeleted = true;
        } else {
            AVLnode* temp = findmin(node->right);
            node->val = temp->val;
            node->right = remove(node->right, temp->val, isDeleted);
        }
    }
    if (node == nullptr) return node;
    node->height = 1 + std::max(height(node->left), height(node->right));
    int balance = getBalance(node);
    if (balance > 1 && getBalance(node->left) >= 0) {
        return rightRotate(node);
    }
    if (balance > 1 && getBalance(node->left) < 0) {
        node->left = leftRotate(node->left);
        return rightRotate(node);
    }
    if (balance < -1 && getBalance(node->right) <= 0) {
        return leftRotate(node);
    }
    if (balance < -1 && getBalance(node->right) > 0) {
        node->right = rightRotate(node->right);
        return leftRotate(node);
    }
    return node;
}


bool AVL::remove(int val) {
    bool isDeleted = false;
    root = remove(root, val, isDeleted);
    return isDeleted;
}
